src/sensors/distant.cpp

```diff
@@ -115,23 +234,29 @@ MTS_VARIANT class DistantSensor final : public Sensor<Float, Spectrum> {
         ray.d      = trafo.transform_affine(Vector3f{ 0.f, 0.f, 1.f });
 
         // 3. Sample ray origin
-        if (!m_has_target) {
-            // If no target point is defined, sample a target point on the
+        if constexpr (OriginType == RayOriginType::Distant) {
+            // If no origin is defined, sample a target point on the
             // bounding sphere's cross section
             Point2f offset =
                 warp::square_to_uniform_disk_concentric(aperture_sample);
             Vector3f perp_offset =
                 trafo.transform_affine(Vector3f{ offset.x(), offset.y(), 0.f });
             ray.o = m_bsphere.center + (perp_offset - ray.d) * m_bsphere.radius;
-        } else {
-            ray.o = m_target - 2.f * ray.d * m_bsphere.radius;
+        } 
+        if constexpr (OriginType == RayOriginType::Rectangle) {
+            Float origin_x = abs(m_origin_b.x() - m_origin_a.x()) * aperture_sample.x() + min(m_origin_a.x(), m_origin_b.x());
+            Float origin_y = abs(m_origin_b.y() - m_origin_a.y()) * aperture_sample.y() + min(m_origin_a.y(), m_origin_b.y());
+            ray.o = Point3f{origin_x, origin_y, m_origin_a.z()};
+        }
+        if constexpr (OriginType == RayOriginType::Disk) {
+            Point2f disk_sample = Point2f(warp::square_to_uniform_disk(aperture_sample)) * m_origin_radius;
+            Point3f sample_disk = Point3f(disk_sample.x(), disk_sample.y(), 0.f) + m_origin_center;
+            ray.o = Point3f{sample_disk.x(), sample_disk.y(), m_origin_center.z()};
         }
 
         ray.update();
         return std::make_pair(
-            ray, m_has_target
-                     ? wav_weight
-                     : wav_weight * (math::Pi<Float> * sqr(m_bsphere.radius)));
+            ray, wav_weight * m_origin_area * Frame3f::cos_theta(-ray.d));
```

Comment:
Ray weights will be different for each ray origin case: you'll need to compute them as part of your `constrexpr if`s.

- For `Distant`: keep the original value.
- For `Rectangle` and  `Disk`: apply the slanting factor.

Reply:
I am not sure about this. My implementation uses the cross section of the bounding sphere that lies in the xy-plane, so I would expect the slanting factor to be necessary.